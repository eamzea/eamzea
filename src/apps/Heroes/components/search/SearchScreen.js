import React, { useMemo } from "react";
import { Row, Col, Form, Button } from "react-bootstrap";
import HeroCard from "../heroes/HeroCard";
import { useForm } from "../../hooks/useForm";
import { useLocation } from "react-router-dom";
import queryString from "query-string";
import { getHeroByName } from "../../selectors/getHeroByName";

const SearchScreen = ({ history }) => {
  const location = useLocation();

  let { q: query } = queryString.parse(location.search);

  if (!query) {
    query = "";
  }

  const [formValues, handleInputChange] = useForm({
    searchText: query ? query : "",
  });

  const { searchText } = formValues;

  const heroesFiltered = useMemo(() => getHeroByName(query), [query]);

  const handleSearch = (e) => {
    e.preventDefault();

    history.push(`?q=${searchText}`);
  };

  return (
    <div>
      <Row>
        <Col xs={12}>
          <h1>Search Screen</h1>
        </Col>
      </Row>
      <Row>
        <Col xs={5}>
          <p className="h4">Search Form</p>
          <hr />
          <Form onSubmit={handleSearch}>
            <Row>
              <Col xs={10}>
                <Form.Control
                  type="text"
                  placeholder="Find your Hero"
                  name="searchText"
                  value={searchText}
                  onChange={handleInputChange}
                />
              </Col>
              <Col xs={2}>
                <Button variant="outline-primary" type="submit">
                  Search
                </Button>
              </Col>
            </Row>
          </Form>
        </Col>
        <Col xs={7}>
          <p className="h4">Results</p>
          <hr />

          {query === "" && (
            <div className="p-3 mb-2 bg-info text-white rounded text-center">
              <p className="h3">Search a hero </p>
            </div>
          )}

          {query !== "" && heroesFiltered.length === 0 && (
            <div className="p-3 mb-2 bg-info text-white rounded text-center">
              <p className="h3">There is no a hero called </p>
              <p className="h4">{query}</p>
            </div>
          )}

          {heroesFiltered.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </Col>
      </Row>
    </div>
  );
};

export default SearchScreen;
