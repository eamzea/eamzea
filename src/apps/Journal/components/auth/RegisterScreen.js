import React from 'react';
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';
import { setError, removeError } from '../../actions/ui';
import { startRegisterEmailPasswordName } from '../../actions/auth';

const RegisterScreen = () => {
  const dispatch = useDispatch();

  const { msgError } = useSelector(state => state.ui);

  const [formValues, handleInputChange] = useForm({
    name: 'Edgar',
    email: 'eamzea@gmail.com',
    password: 'cesarZEA210394',
    password2: 'cesarZEA210394',
  });

  const { name, email, password, password2 } = formValues;

  const handleRegister = e => {
    e.preventDefault();

    if (isFormValid()) {
      dispatch(startRegisterEmailPasswordName(email, password, name));
    }
  };

  const isFormValid = () => {
    if (name.trim().length === 0) {
      dispatch(setError('Name is Required'));
      return false;
    } else if (!validator.isEmail(email)) {
      dispatch(setError('Email is not valid'));
      return false;
    } else if (password.length <= 5) {
      dispatch(setError('Password should be at least 6 characters'));
      return false;
    } else if (password !== password2) {
      dispatch(setError(`Passwords don't match`));
      return false;
    }

    dispatch(removeError());

    return true;
  };

  return (
    <>
      <h3 className="auth__title">Register</h3>
      <form
        onSubmit={handleRegister}
        className="animate__animated animate__fadeIn animate__faster"
      >
        {msgError && <div className="auth__alert-error">{msgError}</div>}

        <input
          className="auth__input"
          type="text"
          placeholder="Name"
          name="name"
          autoComplete="off"
          value={name}
          onChange={handleInputChange}
        />
        <input
          className="auth__input"
          type="text"
          placeholder="Email"
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          className="auth__input"
          type="password"
          placeholder="Password"
          name="password"
          value={password}
          onChange={handleInputChange}
        />
        <input
          className="auth__input"
          type="password"
          placeholder="Confirm password"
          name="password2"
          value={password2}
          onChange={handleInputChange}
        />
        <button type="submit" className="btn btn-primary btn-block mb-5">
          Register
        </button>

        <Link to="/journal/auth/login" className="link">
          Already registered?
        </Link>
      </form>
    </>
  );
};

export default RegisterScreen;
