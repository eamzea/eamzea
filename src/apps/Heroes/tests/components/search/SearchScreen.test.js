import "@testing-library/jest-dom";
import React from "react";
import { mount } from "enzyme";
import { AuthContext } from "../../../auth/AuthContext";
import SearchScreen from "../../../components/search/SearchScreen";
import { MemoryRouter, Route } from "react-router-dom";

describe("Pruebas en <SearchScreen/>", () => {
  test("Debe mostrarse correctamente con valores por defecto", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".h3").text().trim()).toBe("Search a hero");
  });

  test("Debe mostrar a Batman y el input con el valor del query string", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find("input").prop("value")).toBe("batman");
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe mostrar error si no encuentra hero", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman12345"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find(".h3").text().trim()).toBe("There is no a hero called");
    expect(wrapper.find(".bg-info").childAt(1).text().trim()).toBe(
      "batman12345"
    );
  });

  test("Debe llamar push del history", () => {
    const historyMock = {
      push: jest.fn(),
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman12345"]}>
        <Route
          path="/search"
          component={() => <SearchScreen history={historyMock} />}
        />
      </MemoryRouter>
    );

    wrapper.find("input").simulate("change", {
      target: {
        name: "searchText",
        value: "batman",
      },
    });

    wrapper.find("form").prop("onSubmit")({
      preventDefault() {},
    });

    expect(historyMock.push).toHaveBeenCalledWith(`?q=batman`);
  });
});
