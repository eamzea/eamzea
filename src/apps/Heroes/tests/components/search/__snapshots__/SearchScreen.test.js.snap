// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen/> Debe mostrar a Batman y el input con el valor del query string 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-12"
    >
      <h1>
        Search Screen
      </h1>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <p
        className="h4"
      >
        Search Form
      </p>
      <hr />
      <form
        className=""
        onSubmit={[Function]}
      >
        <div
          className="row"
        >
          <div
            className="col-10"
          >
            <input
              className="form-control"
              name="searchText"
              onChange={[Function]}
              placeholder="Find your Hero"
              type="text"
              value="batman"
            />
          </div>
          <div
            className="col-2"
          >
            <button
              className="btn btn-outline-primary"
              disabled={false}
              type="submit"
            >
              Search
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <p
        className="h4"
      >
        Results
      </p>
      <hr />
      <div
        className="card"
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-5"
          >
            <img
              alt="Batman"
              className="card-img-top"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-7"
          >
            <div
              className="card-body"
            >
              <div
                className="card-title h5"
              >
                Batman
              </div>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                <button
                  className="btn btn-outline-primary"
                  disabled={false}
                  type="button"
                >
                  Mas...
                </button>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen/> Debe mostrarse correctamente con valores por defecto 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-12"
    >
      <h1>
        Search Screen
      </h1>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <p
        className="h4"
      >
        Search Form
      </p>
      <hr />
      <form
        className=""
        onSubmit={[Function]}
      >
        <div
          className="row"
        >
          <div
            className="col-10"
          >
            <input
              className="form-control"
              name="searchText"
              onChange={[Function]}
              placeholder="Find your Hero"
              type="text"
              value=""
            />
          </div>
          <div
            className="col-2"
          >
            <button
              className="btn btn-outline-primary"
              disabled={false}
              type="submit"
            >
              Search
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <p
        className="h4"
      >
        Results
      </p>
      <hr />
      <div
        className="p-3 mb-2 bg-info text-white rounded text-center"
      >
        <p
          className="h3"
        >
          Search a hero 
        </p>
      </div>
    </div>
  </div>
</div>
`;
